### PR TITLE
Bugfix: key tracking for 'slice()'-method

### LINKED ISSUE #151: slice() can invalidate primary keys
if positive indexes are repeated.

### PR DESCRIPTION
**Pull request by TSchiefer**
- now drops potential primary key if positive indices are repeated
- now only accepts integer and numeric vectors without any decimal place different from `0`

fixes #151

### CODE REVIEW

File: R/dplyr.R [hunk 1]
```diff
@@ -157,8 +157,21 @@ slice.dm <- function(.data, ...) {
 }
 
 #' @export
-slice.zoomed_dm <- function(.data, ...) {
-  replace_zoomed_tbl(.data, slice(get_zoomed_tbl(.data), ...))
+slice.zoomed_dm <- function(.data, idx) {
+  if (is_missing(idx)) return(.data)
+
+  if (!inherits(idx, "integer")) {
```
Comment (krlmlr) [inline R/dplyr.R]: I forgot to mention that we can use `vctrs::vec_as_index()` to create a positive index vectors from all inputs.

File: R/dplyr.R [hunk 2]
```diff
@@ -157,8 +157,21 @@ slice.dm <- function(.data, ...) {
...
+    # numeric vector only accepted if identical to itself as integer
+    if (!inherits(idx, "numeric") || !all(as.integer(idx) == idx)) abort_need_int(class(idx))
+  }
+  # FIXME: will be easier if we have an extra row in `def` for the zoomed table
+  if (cdm_has_pk(.data, !!orig_name_zoomed(.data))) {
```
Comment (krlmlr) [inline R/dplyr.R]: Can we init `tracked_keys <- NULL` here and return `replace_zoomed_tbl(..., tracked_keys)` unconditionally?

File: R/dplyr.R (hunk 1 above)
Comment (TSchiefer) [reply R/dplyr.R]: would that mean that different from `dplyr::slice()` we would accept logical vectors too(?), or should we actively check against that?

Comment (TSchiefer) [reply R/dplyr.R]: I would opt for not accepting logical vectors, since people might get into habits that don't work with `dplyr::slice()`

Comment (TSchiefer) [reply R/dplyr.R]: I kept the `class()`-checks, for reasons mentioned in the comment in line 162

**Comment by TSchiefer** [comment]
Yup, that way we don't lose the usual `dplyr::slice()` functionality, I like that.

I'd say a message in case `.keep_pk = NULL` AND the original zoomed table has a PK (or: in case `.keep_pk = NULL` AND the original zoomed table has a PK AND that PK is still being tracked)

**Pull request merged**